pkg/versioned/map_test.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package versioned
+
+import (
+	"testing"
+
+	"github.com/cilium/cilium/pkg/lock"
+
+	"gopkg.in/check.v1"
+)
+
+// Hook up gocheck into the "go test" runner.
+func Test(t *testing.T) {
+	check.TestingT(t)
+}
+
+type VersionedSuite struct{}
+
+var _ = check.Suite(&VersionedSuite{})
+
+func (s *VersionedSuite) TestComparableMap_AddEqual(c *check.C) {
+	type fields struct {
+		m          Map
+		deepEquals DeepEqualFunc
+	}
+	type args struct {
+		uuid UUID
+		obj  Object
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		want   bool
+	}{
+		{
+			name: "same object inserted twice should return true because the version is the same",
+			fields: fields{
+				m: Map{
+					"foo": {
+						Data:    nil,
+						Version: ParseVersion("1"),
+					},
+				},
+				deepEquals: DeepEqualFunc(func(o1, o2 interface{}) bool {
+					return true
+				}),
+			},
+			args: args{
+				uuid: "foo",
+				obj: Object{
+					Data:    nil,
+					Version: ParseVersion("1"),
+				},
+			},
+			want: true,
+		},
+		{
+			name: "new object inserted should return false",
+			fields: fields{
+				m: Map{},
+				deepEquals: DeepEqualFunc(func(o1, o2 interface{}) bool {
+					return true
+				}),
+			},
+			args: args{
+				uuid: "foo",
+				obj: Object{
+					Data:    nil,
+					Version: ParseVersion("1"),
+				},
+			},
+			want: false,
+		},
+		{
+			name: "new object with new version but deep equaled should return true",
+			fields: fields{
+				m: Map{
+					"foo": {
+						Data:    nil,
+						Version: ParseVersion("1"),
+					},
+				},
+				deepEquals: DeepEqualFunc(func(o1, o2 interface{}) bool {
+					return true
+				}),
+			},
+			args: args{
+				uuid: "foo",
+				obj: Object{
+					Data:    nil,
+					Version: ParseVersion("2"),
+				},
+			},
+			want: true,
+		},
+		{
+			name: "new object with new version but not equaled should return false",
+			fields: fields{
+				m: Map{
+					"foo": {
+						Data:    nil,
+						Version: ParseVersion("1"),
+					},
+				},
+				deepEquals: DeepEqualFunc(func(o1, o2 interface{}) bool {
+					return false
+				}),
+			},
+			args: args{
+				uuid: "foo",
+				obj: Object{
+					Data:    nil,
+					Version: ParseVersion("2"),
+				},
+			},
+			want: false,
```

Comment:
Can you add a test that tests where the object in the map has version 2, but the object being added has version 1?